### DISCUSSION TITLE
No ICE candidates, No connectionStateChange events

### DISCUSSION THREAD

**Discussion by mehrvarz** [ungoogled-software/ungoogled-chromium-android#116] [General]
I have new findings regarding webRTC-policy patch.

There are two major issues. First, with your WebRTC patch, no ICE candidates are being generated. This prevents P2P connections from being established. The failure occurs without an error message or any explanation towards the user. This means that the WebRTC app developer is likely to receive complaints.

The second problem is even more severe. The connectionStateChange [1] event handler is never called. WebRTC apps often depend on this handler to be called at some point. If it is never called, it can cause flawless WebRTC apps to hang (!) without any user feedback. All web browsers call this handler on any OS - including Ungoogled-Chromium on Mac, Windows and Linux. But your implementation on Android does not, causing apps to hang.

Apps hanging for no apparent reason is a really bad experience. This is why I ask you to resolve these issues, for instance by returning error or "disconnect" messages where functionality is being disabled. If you break apps, at least let them fail graciously.

[1] https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/connectionstatechange_event

**Comment by wchen342** [comment]
I read the [original discussion on this](https://github.com/ungoogled-software/ungoogled-chromium-android/issues/14) again and I think the Bromite patch is breaking on purpose for safe fallback. What if you use this patch instead: https://github.com/GrapheneOS/Vanadium/blob/12/patches/0051-most-private-WebRTC-IP-handling-policy-by-default.patch? Will it still break things?

**Reply by mehrvarz** [reply]
It is not about my needs. I prefer to use WebRTC unpatched.

The problem is with others running into this. WebRTC, by nature, is a little difficult to get working. And now on top of it your browser is breaking these apps without any user feedback. Try running any of these apps https://webrtc.github.io/samples/ in Kiwi and then in your browser. In your browser they fail 100% silently, which is pure ramboism.

Afaics they all fail. It would be better to turn WebRTC off (so it will result in a useful error message) and be honest about it.

**Reply by wchen342** [reply]
That is unexpected indeed. The patch was not made by me, I took it from Bromite and I may need to revert it in that case. I was just wondering whether using Ip handling policy will not break the API, unlike the current approach, but yes I can test it myself.

@csagan5 @uazo is this the intended behavior that WebRTC APIs fail? I was expecting only partial breakage of WebRTC w.r.t. local addresses but it seems the breakage is much more extensive.

**Reply by csagan5** [reply]
The Bromite patch enabled extra logging (see `WebRTC routing preferences` log lines) that can be used to identify what is happening; the patch from Vanadium is not equivalent and would still create privacy problems in some cases, see [this comment of mine](https://github.com/bromite/bromite/issues/553#issuecomment-626196348).

Biggest issue is that the preference `kWebRTCIPHandlingPolicy` seems to have no flag, so it is not possible for the user to change it from the restrictive default.

> In your browser they fail 100% silently

They do not fail in Bromite though.

**Reply by mehrvarz** [reply]
> They do not fail in Bromite though.

When you run [Basic peer connection demo in a single tab](https://webrtc.github.io/samples/src/content/peerconnection/pc1/) in Bromite, does it show you two almost identical live camera feeds?

**Reply by uazo** [reply]
> is this the intended behavior that WebRTC APIs fail?

I don't know yet, I need to verify the code

> Biggest issue is that the preference kWebRTCIPHandlingPolicy seems to have no flag, so it is not possible for the user to change it from the restrictive default.

i could add a content settings, i think i can.

> They do not fail in Bromite though.

It doesn't work for me if `UnifiedAutoplay = FEATURE_DISABLED_BY_DEFAULT`

**Reply by csagan5** [reply]
> > They do not fail in Bromite though.
> 
> When you run [Basic peer connection demo in a single tab](https://webrtc.github.io/samples/src/content/peerconnection/pc1/) in Bromite, does it show you two almost identical live camera feeds?

I have success only with the audio ones; for the cameras I see black squares. I think @uazo might have already found the reason for that.



> > Biggest issue is that the preference kWebRTCIPHandlingPolicy seems to have no flag, so it is not possible for the user to change it from the restrictive default.
> 
> i could add a content settings, i think i can.

That would be even better than the `chrome://flags`, although a more messy patch to maintain.

> > They do not fail in Bromite though.
> 
> It doesn't work for me if `UnifiedAutoplay = FEATURE_DISABLED_BY_DEFAULT`

This is because with `UnifiedAutoplay = FEATURE_DISABLED_BY_DEFAULT` an interaction is necessary before the camera feed is started? We could create a customized WebRTC test that has the interaction and thus it would work?

By the way, this is another setting that user cannot control, similar to the WebRTC one.

**Reply by mehrvarz** [reply]
> I have success only with the audio ones; for the cameras I see black squares. I think @uazo might have already found the reason for that.

[Web Audio output as input to peer connection](https://webrtc.github.io/samples/src/content/peerconnection/webaudio-input/) is audio-only and it does not work in UCA and Bromite. If it works, you can hear yourself talk with a delay (or double if you tick on "Add local..."). The delay comes from the data being sent over the network before playback "on the other side". UnifiedAutoplay is not the (main) culprit. enable_multiple_routes and enable_nonproxied_udp are, since they are disabled.

**Reply by uazo** [reply]
> I think @uazo might have already found the reason for that.

yes, I'm looking at it, but it seems to me that it can only be solved by changing the policy.
to get a more technical answer you would have to investigate the code, but it is quite complicated and perhaps not worth it.
[here](https://groups.google.com/g/discuss-webrtc/c/4Yggl6ZzqZk/m/nV24XkXXAQAJ) some more info

> although a more messy patch to maintain.

I'm working on this with [this solution](https://github.com/uazo/bromite/blob/master-v98/build/patches/Content-settings-infrastructure.patch). it would allow me to simply add content settings and simplify rebases. is part of those things that I would like to propose in bromite (with an [example](https://github.com/uazo/bromite/blob/master-v98/build/patches/Add-site-settings-for-webgl.patch) of use, with this it is [easier to read](https://uazo.github.io/superpatch/bromite/commit/9578f8ce91fdc4d639b8e3cef7bad90e1123326a)). 

> an interaction is necessary before the camera feed is started?

no, it doesn't seem enough: the interaction is already there with the click on the button

**Reply by wchen342** [reply]
> it seems to me that it can only be solved by changing the policy.

Maybe just add flags for switching the policies? So that whoever wants them switched on can do it (For both autoplay and webrtc). Flags are also easier to add then Java settings (they have less lines).

**Reply by uazo** [reply]
> Maybe just add flags for switching the policies?

sure, it's an alternative.
however from what I understand the private address is only necessary for some sites. with the content settings I could always expose the public one only (or have it deactivated) and activate it only for the sites I need (and trust) without having to restart.

**Reply by csagan5** [reply]
I think we should try different policy and settings combinations and use Bromite/UC-Android logs to troubleshoot this until a full picture becomes visible; no changes should be made without testing.

> UnifiedAutoplay is not the (main) culprit. enable_multiple_routes and enable_nonproxied_udp are, since they are disabled.

Have you tested all combinations and reached this conclusion?

**Reply by uazo** [reply]
> Have you tested all combinations and reached this conclusion?

i will try too but reading the code and how `RTCPeerConnection` works, it seems to be like @mehrvarz says. an on/off content settings is more than enough

**Reply by csagan5** [reply]
`enable_multiple_routes` and `enable_nonproxied_udp` are disabled because they would leak the IP-address-behind-VPN; the end result should be that WebRTC can work with or without VPN and without leaking any other IP address than the public one.

So an on/off content setting is not enough, despite the oversimplified description of the problem in this thread; it should be established (via tests as mentioned before) if WebRTC can be made to work without leaking such IP address, and then that can become the default in Bromite.

**Reply by uazo** [reply]
surely you know more than me, but I'll try and then we'll talk about it

**Reply by mehrvarz** [reply]
> despite the oversimplified description of the problem in this thread

*irrefutable

**Reply by csagan5** [reply]
The issue was summarised here: https://github.com/bromite/bromite/issues/553#issuecomment-626196348

CCTs would be able to expose the local IP address (without VPN) with the default policy; thus the default policy is patches in Bromite to not allow `multiple_routes` and nonproxied_udp`.

**Reply by wchen342** [reply]
I think the concern here is not the IP, instead the concern is that because the switch is not standard implementation, JavaScript codes fail silently when running, thus make applications unusable without any indication.

**Reply by csagan5** [reply]
> I think the concern here is not the IP, instead the concern is that because the switch is not standard implementation, JavaScript codes fail silently when running, thus make applications unusable without any indication.

If no WebRTC can be established the Javascript bindings should return an error; if they do not, that is a bug to fix.

**Comment by mehrvarz** [comment]
Trash the policy. Stop doctoring. It cannot work because no one wants a crippled WebRTC. Instead, allow users to turn it off completely.

[ x ] Disable WebRTC

Make it enabled by default, like a conformant browser. And tell VPN users they need to check this ON to prevent IP leakage.

In disabled state I expect RTCPeerConnection() to return null (Edit: or NotFoundError). This will ensure WebRTC apps don't "appear to work" while being hung with no user feedback.

**Reply by mehrvarz** [reply]
### Why a static policy is not a solution

What enables WebRTC apps to connect to each other in p2p fashion and what is experienced by VPN users as ip-leakage is the exact same thing. It is the ability of Javascript code to find out the public ip-address of the device it is running on. If your device cannot find out it's current public ip-address and therefore cannot share this info with others (via signaling server), no low-lateny p2p connection can be established. This is why the privacy concern cannot be solved with a static policy. Users must be able to express their intent. A switch is required.

### Why deactivating only parts of WebRTC is not a solution

When a user disables WebRTC (aka enables ip-leak protection), this should not lead to misbehaving WebRTC apps. Instead, WebRTC apps should fail quickly and visibly. The best way to achieve this is to disable WebRTC completely, not just parts of it. WebRTC apps exiting with an error message (NotFoundError) right on start will give the user a good idea as to what the problem is and how it can be solved. Merely removing ICE candidates makes apps hang with no user feedback. Users will not understand what the problem is, even if they have deactivated WebRTC just minutes earlier.

**Reply by wchen342** [reply]
I went back to check how Firefox did the WebRTC leak prevention, and it turns out on Firefox you simply disable WebRTC, so yes, probably disabling is better. The whole problem actually started from there is no flag to disable WebRTC on Chromium-based browsers.
Brave has a [more complicated approach](https://support.brave.com/hc/en-us/articles/360017989132-How-do-I-change-my-Privacy-Settings-#webrtc) but it doesn't seems to be easily achievable on Android because of the interface thing.
In this case I now agree with you that a switch to disable WebRTC is better.

**Reply by mehrvarz** [reply]
> there is no flag to disable WebRTC on Chromium-based browsers.

There used to be one. Do you think it would make sense to check out an old version of Chromium to see how it was implemented?

Do you know what tools/metrics/histograms/enums.xml is good for? There is an enum 563980787 in there referring to "disable-webrtc".

**Reply by wchen342** [reply]
That is just usage statistics. Chromium has calls to gather usage data and save them in a lot of its internal functions. The file is just a list of entries they collect for usage statistics.

A flag to disable WebRTC completely shouldn't be hard to do. We just need to make sure (1) it is really disabled and (2) JS gives the correctly exception. That is doable I think.

**Reply by mehrvarz** [reply]
> turns out on Firefox you simply disable WebRTC

You are referring to: media.peerconnection.enabled. After I set this to false, new RTCPeerConnection() in JS throws an exception: "ReferenceError: RTCPeerConnection is not defined". This is easy to catch.

> A flag to disable WebRTC completely shouldn't be hard to do.

Appreciate your optimism. Please keep me updated on this journey. If I could make one wish: don't bury the switch deep in settings. Put it in the main menu, next to "Desktop site [ ]". So the current state can always be checked quickly.

**Reply by mehrvarz** [reply]
> "ReferenceError: RTCPeerConnection is not defined". This is easy to catch.

What I meant to say: Everybody is likely to catch this.

Here is one of the sample apps referenced above running into WebRTC being disabled in FF:
https://timur.mobi/li/Screenshot-2022-02-21-00-28-56.png
The error message is a little cryptic (talking about "local media"). But the app has visibly stopped acting.

Here is my WebCall app running into WebRTC being disabled in FF: 
https://timur.mobi/li/Screenshot-2022-02-21-00-29-44.png

**Comment by uazo** [comment]
after checking the code here is what i found.

there are three options:
- **enable_multiple_routes**
_true (default)_: if the page has access rights to the camera or microphone, it shows the local ip, otherwise only the public one. this logic is also applied to datachannels
_false_: access to the ip is regardless of the permissions granted and controlled by enable_default_local_candidate

- **enable_nonproxied_udp**
_true (default)_: Enable both udp and tcp
_false_: disable transmission via udp (so keep tcp only)

- **enable_default_local_candidate**
_true (default)_: enable transmission of local ip
_false_: disable the transmission of local ip

the logic of enable_multiple_routes has priority over enable_default_local_candidate, the latter is not taken into account if enable_multiple_routes = true

@csagan5  I have not noticed any differences in the management of the cct, neither application nor functional. Would you give me a way to reproduce what you say?

about the error condition: the documentation explicitly says

```
RTCPeerConnection.onicecandidate:
If the event's candidate property is null, ICE gathering has finished.
https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/onicecandidate
```

This is not actually true in chromium, because due to the filtering done on candidates instead of deleting the value, the passing of null is one per interface.

But the documentation also says:

```
You don't need to watch for this explicitly; instead, if you need to sense the end of signaling, you should watch for a
icegatheringstatechange event indicating that the ICE negotiation has transitioned to the complete state.
https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/onicecandidate
```

so it seems to me that there is already a way to know if the webrtc negotiation cannot be successful by verifying the lack of ice candidates or the status of the event.

regarding the possibility of throwing an exception to the creation of RTCPeerConnection, although it does not seem foreseen in the
w3c documentation (https://www.w3.org/TR/webrtc/#constructor) but technically we would only break sites that do not comply with the specifications (i.e. without a try / catch) because in the idl it is contemplated that it might throw an exception (has RaisesException).

ultimately, we could go back to default and remove the patch, after all having webrtc only active on the public interface i think has
little sense in most cases while the IP leak seems mandatory given the p2p nature of the transmission and the user must grant the
permission for the camera or the microphone (which is perhaps worse than granting the ip)
there has also been an attempt to use mDNS (active by default) but it seems that the library used is not able to handle it (for now) because it requires a defined ip and not a name.

alternatively, as I said, it is possible to activate a content setting that allows a choice between (EDIT) ~true~ false / true / true (ice candidates provided by the browser) and false / false / false (ice candidates to pass manually, i.e. the site must provide them via javascript), but it would be a specific condition of bromite, perhaps only necessary to allow to activate a datachannel without necessarily granting access to the camera or microphone.

**Reply by csagan5** [reply]
> @csagan5 I have not noticed any differences in the management of the cct, neither application nor functional. Would you give me a way to reproduce what you say?

You could reproduce the tests done here: https://github.com/bromite/bromite/issues/553#issuecomment-626196348

The problem is that CCTs use the default policy, user has no control over that (that is why default policy was changed in the patch); as for Bromite we have the following:
1. there should not be any mode possible that when enabled leaks the local IP address (whether CCT or not, or VPN or not, does not matter)
2. we can discuss a feature to add flags/settings, but (1) should still hold true

We can discuss the above in a Bromite feature request issue.

**Reply by mehrvarz** [reply]
> there should not be any mode possible that when enabled leaks the local IP address ...VPN or not, does not matter

Say I am not using a VPN and I did not disable WebRTC bc I actually want to run some WebRTC apps. The web server I am going to can see my true IP the whole time. Are you sure you want to prevent ICE candidates from working as required in such a case? 
This would prevent P2P links with E2E encryption from being established.

**Reply by csagan5** [reply]
Your Android device has these IP addresses when using WiFi:
* `140.82.121.3` your public IP address
* `192.168.1.50` your local IP address <-- this is the one that should not be leaked in any case, it is useful for fingerprinting and other type of attacks
* `127.0.0.1` localhost address

**Reply by mehrvarz** [reply]
So a browser provides a "[ _ ] Disable WebRTC" feature. Are you saying 192.168.1.50 should be prevented from leaking in cases where WebRTC is NOT disabled?

**Reply by mehrvarz** [reply]
When WebRTC is enabled, neither the public nor the local IP address should be blocked.

With respect to the local IP address, please read [this post](https://github.com/ungoogled-software/ungoogled-chromium-android/discussions/114#discussioncomment-2062035) from end of January. In it I describe how the local IP address can provide very powerful functionality: the dynamic handover of a live P2P connection from wifi to mobile internet (in this particular example). 

WebRTC is not a surveillance tool. Everything in it provides useful and honest functionality. It is not like a tracker inside a web page that you can just root out. If there is a WebRTC kill switch avail to the user, there is no need to modify the way WebRTC works. If you come to a different conclusion, how about disabling WebRTC completely?

Here is what I would find useful: a visual feedback for the user, indicating the mode the browser is in. Say, a red dot close to (or inside) the URL box, that lights up when WebRTC is enabled. This would prevent the user from making a false assumption.

**Reply by csagan5** [reply]
I will not continue discussing a Bromite patch here; feel free to change the patch as it best fits your use.

**Reply by wchen342** [reply]
Yes, I want to point out that although I borrow patches from Bromite, ungoogled-chromium and Bromite are different projects and have different goals. I will go ahead and implement the switch approach, while @csagan5 is free to do what he see fits for Bromite.

**Reply by csagan5** [reply]
I made a summary of webRTC in Bromite here: https://github.com/bromite/bromite/wiki/WebRTC

It also mentions what is the information leaked by default so that @wchen342 you can decide the strategy based on that.

**Comment by wchen342** [comment]
I have changed the way WebRTC is handled. A new flag is added and when it is enabled, `new RTCPeerConnection()` will now throw a [DOMException](https://developer.mozilla.org/en-US/docs/Web/API/DOMException/DOMException) with `NotSupportedError`.
A per-site UI is planned, but I need to get the update first so it will be added later.

**Reply by mehrvarz** [reply]
There is an app in F-Droid that makes verifying the System WebView super easy. It's called WebRTC-Check and it can be found here: https://f-droid.org/packages/timur.webrtc.check

**Reply by uazo** [reply]
> A per-site UI is planned, but I need to get the update first so it will be added later.

I've already seen how to do it. if you need just tell me

**Reply by csagan5** [reply]
I have opened a Bromite feature request for this: https://github.com/bromite/bromite/issues/1965
